File: src/AgentOrchestrator/Agent/OrchestratorAgent.cs
using System.Text.Json;
using AgentOrchestrator.Auth;
using AgentOrchestrator.Constants;
using AgentOrchestrator.Models;
using Microsoft.Agents.Builder;
using Microsoft.Agents.Builder.App;
using Microsoft.Agents.Builder.State;
using Microsoft.Agents.Core.Models;
using Microsoft.SemanticKernel;

namespace AgentOrchestrator.Agent;

public class OrchestratorAgent : AgentApplication
{
    private readonly Kernel _kernel;
    private readonly ITokenService _tokenService;
    private readonly IHttpContextAccessor _httpContextAccessor;
    private readonly OrchestrationSettings _orchestrationSettings;
    private readonly ILogger<OrchestratorAgent> _logger;

    private const int MaxMessageLength = 4000;
    
    /// <summary>
    /// OAuth handler name configured in appsettings.json under AgentApplication:UserAuthorization:Handlers
    /// </summary>
    private const string OAuthHandlerName = "graph";

    public OrchestratorAgent(
        AgentApplicationOptions options,
        Kernel kernel,
        ITokenService tokenService,
        IHttpContextAccessor httpContextAccessor,
        OrchestrationSettings orchestrationSettings,
        ILogger<OrchestratorAgent> logger) : base(options)
    {
        _kernel = kernel;
        _tokenService = tokenService;
        _httpContextAccessor = httpContextAccessor;
        _orchestrationSettings = orchestrationSettings;
        _logger = logger;

        // Register activity handlers
        // Note: We don't use autoSignInHandlers here because we need to handle auth differently
        // for general knowledge queries (no auth needed) vs M365 queries (auth required)
        OnActivity(ActivityTypes.Message, OnMessageActivityAsync);
        OnActivity(ActivityTypes.ConversationUpdate, OnConversationUpdateAsync);
    }

    private async Task OnMessageActivityAsync(
        ITurnContext turnContext,
        ITurnState turnState,
        CancellationToken cancellationToken)
    {
        var activity = turnContext.Activity;
        var userMessage = activity.Text?.Trim() ?? string.Empty;
        
        _logger.LogInformation("=== MESSAGE RECEIVED ===");
        _logger.LogInformation("Channel: {Channel}", activity.ChannelId);
        _logger.LogInformation("From: {From}", activity.From?.Id);
        _logger.LogInformation("Text: {Text}", userMessage);

        if (string.IsNullOrEmpty(userMessage))
        {
            await turnContext.SendActivityAsync(
                MessageFactory.Text("Please enter a message."),
                cancellationToken);
            return;
        }

        // Input validation
        if (userMessage.Length > MaxMessageLength)
        {
            await turnContext.SendActivityAsync(
                MessageFactory.Text($"Message too long. Maximum {MaxMessageLength} characters allowed."),
                cancellationToken);
            return;
        }

        _logger.LogInformation("Processing message: {Message}", userMessage);

        try
        {
            // Create timeout-aware cancellation token
            using var timeoutCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
            timeoutCts.CancelAfter(TimeSpan.FromSeconds(_orchestrationSettings.TimeoutSeconds));
            var timeoutToken = timeoutCts.Token;

            // Step 1: Analyze intent (before auth so we can allow general queries)
            _logger.LogInformation("Step 1: Analyzing intent...");
            var intents = await AnalyzeIntentAsync(userMessage, timeoutToken);

            // Apply MaxAgentCalls limit
            if (intents.Count > _orchestrationSettings.MaxAgentCalls)
            {
                _logger.LogWarning("Truncating intents from {Count} to {Max}",
                    intents.Count, _orchestrationSettings.MaxAgentCalls);
                intents = intents.Take(_orchestrationSettings.MaxAgentCalls).ToList();
            }

            _logger.LogInformation("Detected {Count} intent(s): {Intents}",
                intents.Count,
                string.Join(", ", intents.Select(i => i.Type)));

            // Check if any M365 intents require authentication
            var hasM365Intent = intents.Any(i =>
                i.Type == IntentType.M365Email ||
                i.Type == IntentType.M365Calendar ||
                i.Type == IntentType.M365Files ||
                i.Type == IntentType.M365People);

            // Get user's access token only if needed for M365 calls
            string? accessToken = null;
            if (hasM365Intent)
            {
                try
                {
                    accessToken = await GetUserAccessTokenAsync(turnContext, timeoutToken);
                }
                catch (UnauthorizedAccessException ex)
                {
                    _logger.LogWarning(ex, "Auth failed for M365 intent, falling back to general knowledge");

                    // For Teams, give a helpful message and handle general knowledge intents only
                    var channelId = turnContext.Activity.ChannelId;
                    if (channelId == "msteams")
                    {
                        // Filter to only general knowledge intents
                        var generalIntents = intents.Where(i => i.Type == IntentType.GeneralKnowledge).ToList();

                        if (generalIntents.Count > 0)
                        {
                            // Continue with just general knowledge
                            intents = generalIntents;
                            accessToken = string.Empty; // Not needed for general knowledge
                        }
                        else
                        {
                            // All intents require M365 - user needs to sign in
                            // This should rarely happen since AutoSignIn is enabled
                            await turnContext.SendActivityAsync(
                                MessageFactory.Text(
                                    "To access your M365 data (emails, calendar, files), I need you to sign in first.\n\n" +
                                    "Please click the Sign In button when prompted, or try sending your message again.\n\n" +
                                    "In the meantime, I can answer general knowledge questions. Try:\n" +
                                    "- \"What is machine learning?\"\n" +
                                    "- \"Explain microservices\""),
                                cancellationToken);
                            return;
                        }
                    }
                    else
                    {
                        throw; // Re-throw for web channel
                    }
                }
            }

            // Step 2: Execute agents based on intents
            _logger.LogInformation("Step 2: Executing agents (parallel={Parallel})...",
                _orchestrationSettings.EnableParallelExecution);
            var responses = await ExecuteAgentsAsync(intents, accessToken ?? string.Empty, timeoutToken);

            // Step 3: Synthesize response
            _logger.LogInformation("Step 3: Synthesizing response...");
            var finalResponse = await SynthesizeResponseAsync(userMessage, responses, timeoutToken);

            // Step 4: Send response
            await turnContext.SendActivityAsync(
                MessageFactory.Text(finalResponse),
                cancellationToken);

            _logger.LogInformation("Response sent successfully");
        }
        catch (UnauthorizedAccessException)
        {
            await turnContext.SendActivityAsync(
                MessageFactory.Text("Please log in to access M365 features. Visit the web interface to authenticate."),
                cancellationToken);
        }
        catch (OperationCanceledException) when (!cancellationToken.IsCancellationRequested)
        {
            _logger.LogWarning("Request timed out after {Seconds} seconds", _orchestrationSettings.TimeoutSeconds);
            await turnContext.SendActivityAsync(
                MessageFactory.Text("The request timed out. Please try a simpler query or try again later."),
                cancellationToken);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error processing message");
            // Don't expose internal error details to users
            await turnContext.SendActivityAsync(
                MessageFactory.Text("Sorry, an error occurred processing your request. Please try again."),
                cancellationToken);
        }
    }

    private async Task<List<Intent>> AnalyzeIntentAsync(string query, CancellationToken cancellationToken)
    {
        var result = await _kernel.InvokeAsync(
            PluginNames.Intent,
            "AnalyzeIntent",
            new() { ["query"] = query },
            cancellationToken);

        var json = result.GetValue<string>() ?? "[]";

        try
        {
            var intents = JsonSerializer.Deserialize<List<Intent>>(json, new JsonSerializerOptions
            {
                PropertyNameCaseInsensitive = true
            });

            if (intents == null || intents.Count == 0)
            {
                _logger.LogWarning("Intent analysis returned empty results for query: {Query}", query);
                return [new Intent { Type = IntentType.GeneralKnowledge, Query = query }];
            }

            return intents;
        }
        catch (JsonException ex)
        {
            // DEBUGGING TIP: Always log the actual response when JSON parsing fails.
            // This helps identify if the LLM is returning unexpected formats.
            _logger.LogWarning(ex, "Failed to parse intent response. Raw JSON: {Json}. Defaulting to general knowledge.", json);
            return [new Intent { Type = IntentType.GeneralKnowledge, Query = query }];
        }
    }

    private async Task<List<AgentResponse>> ExecuteAgentsAsync(
        List<Intent> intents,
        string accessToken,
        CancellationToken cancellationToken)
    {
        if (_orchestrationSettings.EnableParallelExecution)
        {
            var tasks = intents.Select(intent => ExecuteAgentForIntentAsync(intent, accessToken, cancellationToken));
            var responses = await Task.WhenAll(tasks);
            return responses.ToList();
        }
        else
        {
            // Sequential execution
            var responses = new List<AgentResponse>();
            foreach (var intent in intents)
            {
                var response = await ExecuteAgentForIntentAsync(intent, accessToken, cancellationToken);
                responses.Add(response);
            }
            return responses;
        }
    }

    private async Task<AgentResponse> ExecuteAgentForIntentAsync(
        Intent intent,
        string accessToken,
        CancellationToken cancellationToken)
    {
        try
        {
            return intent.Type switch
            {
                IntentType.M365Email => await ExecuteM365PluginAsync("QueryEmails", intent.Query, accessToken, cancellationToken),
                IntentType.M365Calendar => await ExecuteM365PluginAsync("QueryCalendar", intent.Query, accessToken, cancellationToken),
                IntentType.M365Files => await ExecuteM365PluginAsync("QueryFiles", intent.Query, accessToken, cancellationToken),
                IntentType.M365People => await ExecuteM365PluginAsync("QueryPeople", intent.Query, accessToken, cancellationToken),
                IntentType.GeneralKnowledge => await ExecuteGeneralKnowledgeAsync(intent.Query, cancellationToken),
                _ => new AgentResponse
                {
                    Agent = "unknown",
                    IntentType = intent.Type,
                    Content = "I'm not sure how to handle that request.",
                    Success = false
                }
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error executing agent for intent {IntentType}", intent.Type);
            // SECURITY: Don't expose exception details to users
            return new AgentResponse
            {
                Agent = intent.Type.ToString(),
                IntentType = intent.Type,
                Content = "Sorry, I encountered an issue processing this part of your request.",
                Success = false
            };
        }
    }

    private async Task<AgentResponse> ExecuteM365PluginAsync(
        string functionName,
        string query,
        string accessToken,
        CancellationToken cancellationToken)
    {
        var result = await _kernel.InvokeAsync(
            PluginNames.M365Copilot,
            functionName,
            new()
            {
                ["query"] = query,
                ["accessToken"] = accessToken
            },
            cancellationToken);

        return new AgentResponse
        {
            Agent = "m365_copilot",
            IntentType = functionName switch
            {
                "QueryEmails" => IntentType.M365Email,
                "QueryCalendar" => IntentType.M365Calendar,
                "QueryFiles" => IntentType.M365Files,
                "QueryPeople" => IntentType.M365People,
                _ => IntentType.GeneralKnowledge
            },
            Content = result.GetValue<string>() ?? string.Empty,
            Success = true
        };
    }

    private async Task<AgentResponse> ExecuteGeneralKnowledgeAsync(
        string query,
        CancellationToken cancellationToken)
    {
        var result = await _kernel.InvokeAsync(
            PluginNames.AzureOpenAI,
            "GeneralKnowledge",
            new() { ["query"] = query },
            cancellationToken);

        return new AgentResponse
        {
            Agent = "azure_openai",
            IntentType = IntentType.GeneralKnowledge,
            Content = result.GetValue<string>() ?? string.Empty,
            Success = true
        };
    }

    private async Task<string> SynthesizeResponseAsync(
        string originalQuery,
        List<AgentResponse> responses,
        CancellationToken cancellationToken)
    {
        // If only one successful response, return it directly
        var successfulResponses = responses.Where(r => r.Success).ToList();
        if (successfulResponses.Count == 1)
        {
            return successfulResponses[0].Content;
        }

        if (successfulResponses.Count == 0)
        {
            return "I wasn't able to find an answer to your question.";
        }

        // Multiple responses - synthesize them
        var responsesJson = JsonSerializer.Serialize(successfulResponses.Select(r => new
        {
            agent = r.Agent,
            content = r.Content
        }));

        var result = await _kernel.InvokeAsync(
            PluginNames.Synthesis,
            "Synthesize",
            new()
            {
                ["originalQuery"] = originalQuery,
                ["responses"] = responsesJson
            },
            cancellationToken);

        return result.GetValue<string>() ?? "I couldn't synthesize a response.";
    }

    private async Task<string> GetUserAccessTokenAsync(
        ITurnContext turnContext,
        CancellationToken cancellationToken)
    {
        var channelId = turnContext.Activity.ChannelId;

        // Handle Teams channel with SSO
        if (channelId == "msteams")
        {
            return await GetTeamsSsoTokenAsync(turnContext, cancellationToken);
        }

        // For web/emulator channel, get token from HTTP context session
        var httpContext = _httpContextAccessor.HttpContext;
        if (httpContext == null)
        {
            throw new UnauthorizedAccessException("No HTTP context available. Please log in via the web interface.");
        }

        var sessionId = httpContext.Session.Id;
        if (string.IsNullOrEmpty(sessionId))
        {
            throw new UnauthorizedAccessException("No session available. Please log in.");
        }

        try
        {
            var token = await _tokenService.GetAccessTokenAsync(sessionId);
            if (!string.IsNullOrEmpty(token))
            {
                return token;
            }
        }
        catch (InvalidOperationException ex)
        {
            _logger.LogWarning(ex, "Token not found for session {SessionId}", sessionId);
        }

        throw new UnauthorizedAccessException("No access token available. Please log in.");
    }

    /// <summary>
    /// Get access token for Teams users using the M365 Agents SDK UserAuthorization.
    ///
    /// The SDK handles:
    /// 1. Automatic sign-in card display when no token is available
    /// 2. Token caching and refresh
    /// 3. SSO token exchange (when configured)
    ///
    /// Configuration is in appsettings.json under AgentApplication:UserAuthorization
    /// See: https://learn.microsoft.com/en-us/microsoft-365/agents-sdk/agent-oauth-configuration-dotnet
    /// </summary>
    private async Task<string> GetTeamsSsoTokenAsync(
        ITurnContext turnContext,
        CancellationToken cancellationToken)
    {
        var activity = turnContext.Activity;
        var userId = activity.From?.Id ?? "unknown";

        _logger.LogInformation("Getting token for Teams user {UserId} via SDK UserAuthorization", userId);

        try
        {
            // Use the SDK's built-in UserAuthorization to get token
            // This automatically handles sign-in cards, token exchange, and caching
            var token = await UserAuthorization.GetTurnTokenAsync(turnContext, OAuthHandlerName);
            
            if (!string.IsNullOrEmpty(token))
            {
                _logger.LogInformation("Token retrieved successfully for Teams user {UserId}", userId);
                return token;
            }
            
            _logger.LogWarning("GetTurnTokenAsync returned null/empty for user {UserId}", userId);
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Failed to get token via SDK for user {UserId}", userId);
        }

        // If SDK didn't return a token, sign-in is required
        // The SDK should have already sent a sign-in card, but we'll throw to stop processing
        throw new UnauthorizedAccessException(
            "Please sign in using the button above to access your M365 data.");
    }

    private async Task OnConversationUpdateAsync(
        ITurnContext turnContext,
        ITurnState turnState,
        CancellationToken cancellationToken)
    {
        var membersAdded = turnContext.Activity.MembersAdded;
        if (membersAdded != null)
        {
            foreach (var member in membersAdded)
            {
                if (member.Id != turnContext.Activity.Recipient.Id)
                {
                    await turnContext.SendActivityAsync(
                        MessageFactory.Text(
                            "Welcome to the .NET 10 Agent! I can help you with M365 data (emails, calendar, files, people) " +
                            "and general knowledge questions. Try asking something like:\n\n" +
                            "- \"Summarize my emails from today\"\n" +
                            "- \"What meetings do I have tomorrow?\"\n" +
                            "- \"Explain what microservices are\""),
                        cancellationToken);
                }
            }
        }
    }
}

File: Python/src/agent/orchestrator_agent.py
"""
Orchestrator agent using M365 Agents SDK AgentApplication pattern.

This follows the same pattern as the .NET OrchestratorAgent.cs which extends
AgentApplication and uses OnActivity handlers for message routing.
"""

import asyncio
import json
import uuid
from typing import List, Optional
import logging

from microsoft_agents.hosting.core import (
    Authorization,
    TurnContext,
    MessageFactory,
    MemoryStorage,
    AgentApplication,
    TurnState,
    MemoryStorage,
)
from microsoft_agents.activity import load_configuration_from_env, ActivityTypes
from microsoft_agents.authentication.msal import MsalConnectionManager
from microsoft_agents.hosting.fastapi import CloudAdapter

from semantic_kernel import Kernel
from semantic_kernel.connectors.ai.open_ai import AzureChatCompletion

from models.intent import Intent, IntentType
from models.agent_response import AgentResponse
from models.configuration import AppSettings
from constants.plugin_names import PluginNames
from plugins.agent_context import AgentContext
from plugins.intent_plugin import IntentPlugin
from plugins.m365_copilot_plugin import M365CopilotPlugin
from plugins.azure_openai_plugin import AzureOpenAIPlugin
from plugins.synthesis_plugin import SynthesisPlugin


# Constants matching .NET implementation
MAX_MESSAGE_LENGTH = 4000
AGENTIC_AUTH_HANDLER = "agentic"
NON_AGENTIC_AUTH_HANDLER = "me"

class OrchestratorAgentApp:
    """
    Orchestrator agent application following M365 Agents SDK pattern.

    This class mirrors the .NET OrchestratorAgent which extends AgentApplication.
    It initializes the agent application, registers handlers, and manages the
    3-step orchestration flow: Intent Analysis → Agent Execution → Response Synthesis.
    """
    

    def __init__(self, settings: AppSettings, connection_manager: MsalConnectionManager):
        """
        Initialize the orchestrator agent application.

        Args:
            settings: Application configuration.
            connection_manager: MsalConnectionManager instance.
        """
        self._settings = settings
        self._logger = logging.getLogger(__name__)

        self._local_connection_manager: MsalConnectionManager = connection_manager

        # Initialize Semantic Kernel with Azure OpenAI
        self._base_kernel = self._create_kernel()

        # Initialize M365 Agents SDK AgentApplication
        self._agent_app = AgentApplication[TurnState](
            storage=MemoryStorage(),
            adapter=CloudAdapter(connection_manager=connection_manager),
            authorization=Authorization(MemoryStorage(),
                            connection_manager= connection_manager, 
                            **settings.agents_sdk_config)
        )

        # Register activity handlers
        self._register_handlers()

        self._logger.info("OrchestratorAgent initialized with M365 Agents SDK pattern")

    @property
    def agent_app(self) -> AgentApplication:
        """Get the underlying AgentApplication instance."""
        return self._agent_app

    def _register_handlers(self):
        """
        Register activity handlers with the agent application.
        """

        # Handle conversation updates (e.g., user joins)
        @self._agent_app.conversation_update("membersAdded")
        async def on_conversation_update(context: TurnContext, state: TurnState):
            """Handle conversation updates when members are added."""
            await context.send_activity(
                "Welcome! I'm a multi-agent orchestrator that can help with "
                "Microsoft 365 data (emails, calendar, files, people) and general knowledge. "
                "Ask me anything!"
            )

        # Handle message activities (user messages)
        # This is the main handler that mirrors OnMessageActivityAsync in .NET (line 47-136)
        @self._agent_app.activity("message", auth_handlers=["GRAPH"])
        async def on_message_activity(context: TurnContext, state: TurnState):
            """
            Handle incoming message activities with 3-step orchestration flow.

            This is the main orchestration handler that mirrors the .NET
            OnMessageActivityAsync method in OrchestratorAgent.cs:47-136.

            Args:
                context: Turn context from M365 Agents SDK.
                state: Turn state for conversation management.
            """
            user_message = context.activity.text.strip() if context.activity.text else ""

            # Input validation (mirrors lines 58-73 in .NET)
            if not user_message:
                await context.send_activity("Please enter a message.")
                return

            if len(user_message) > MAX_MESSAGE_LENGTH:
                await context.send_activity(
                    f"Message too long. Maximum {MAX_MESSAGE_LENGTH} characters allowed."
                )
                return

            self._logger.info(f"Processing message: {user_message}")

            try:
                # Create timeout-aware cancellation (mirrors lines 80-82 in .NET)
                async with asyncio.timeout(self._settings.orchestration.timeout_seconds):
                    # Generate request ID for tracking
                    request_id = str(uuid.uuid4())

                    # Get access token if available
                    # TODO: Implement token acquisition via self._agent_app.auth.get_token()
                    # Mirrors: await _context.UserAuth.ExchangeTurnTokenAsync()
                    access_token = None

                    # Create agent context for this turn
                    agent_context = AgentContext(
                        context=context,
                        state=state,
                        connection_name= 'SERVICE_CONNECTION',
                        user_auth=self._agent_app.auth,
                        auth_handler_name='GRAPH',  #context.authorization.get_active_handler_name(),
                        logger=self._logger
                    )

                    # Setup kernel with plugins for this turn (mirrors line 85)
                    kernel = self._setup_plugins(agent_context)

                    # Step 1: Analyze intent (mirrors lines 87-89)
                    self._logger.info("Step 1: Analyzing intent...")
                    intents = await self._analyze_intent(kernel, user_message)

                    # Apply MaxAgentCalls limit (mirrors lines 92-97)
                    if len(intents) > self._settings.orchestration.max_agent_calls:
                        self._logger.warning(
                            f"Truncating intents from {len(intents)} to "
                            f"{self._settings.orchestration.max_agent_calls}"
                        )
                        intents = intents[:self._settings.orchestration.max_agent_calls]

                    self._logger.info(
                        f"Detected {len(intents)} intent(s): "
                        f"{', '.join(i.type.value for i in intents)}"
                    )

                    # Step 2: Execute agents based on intents (mirrors lines 103-106)
                    self._logger.info(
                        f"Step 2: Executing agents "
                        f"(parallel={self._settings.orchestration.enable_parallel_execution})..."
                    )
                    responses = await self._execute_agents(kernel, intents)

                    # Step 3: Synthesize response (mirrors lines 108-110)
                    self._logger.info("Step 3: Synthesizing response...")
                    final_response = await self._synthesize_response(
                        kernel, user_message, responses
                    )

                    # Step 4: Send response (mirrors line 113)
                    await context.send_activity(final_response)

                    self._logger.info("Response sent successfully")

            except asyncio.TimeoutError:
                # Mirrors lines 121-125 in .NET
                self._logger.warning(
                    f"Request timed out after {self._settings.orchestration.timeout_seconds} seconds"
                )
                await context.send_activity(
                    "The request timed out. Please try a simpler query or try again later."
                )
            except Exception as ex:
                # Mirrors lines 126-131 in .NET
                self._logger.error(f"Error processing message: {ex}", exc_info=True)
                # Don't expose internal error details to users
                await context.send_activity(
                    "Sorry, an error occurred processing your request. Please try again."
                )

        # Register help command
        @self._agent_app.message("/help")
        async def on_help(context: TurnContext, state: TurnState):
            """Handle help command."""
            help_text = """
**Available Commands:**
- Ask questions about your M365 data (emails, calendar, files, people)
- Ask general knowledge questions
- /help - Show this help message

**Example Queries:**
- "What meetings do I have tomorrow?"
- "Summarize my recent emails"
- "What is Docker?"
- "What meetings do I have and what is Kubernetes?" (multi-intent)
"""
            await context.send_activity(help_text)

    def _create_kernel(self) -> Kernel:
        """
        Create base Semantic Kernel instance with Azure OpenAI.

        Returns:
            Configured Kernel instance.
        """
        kernel = Kernel()

        # Add Azure OpenAI chat completion service
        service_id = "azure_openai"
        kernel.add_service(
            AzureChatCompletion(
                service_id=service_id,
                deployment_name=self._settings.azure_openai.deployment_name,
                endpoint=self._settings.azure_openai.endpoint,
                api_key=self._settings.azure_openai.api_key,
                api_version=self._settings.azure_openai.api_version
            )
        )

        return kernel

    def _setup_plugins(self, context: AgentContext) -> Kernel:
        """
        Create per-turn kernel with plugins registered with fresh context.

        This mirrors UpdateChatClientWithToolsInContext in .NET (line 361).
        Plugins are registered per-turn to ensure each has access to current
        turn's context and state.

        Args:
            context: Agent context for this turn.

        Returns:
            Kernel with plugins registered.
        """
        # Clone base kernel for this turn
        kernel = Kernel()

        # Copy service from base kernel
        for service_id, service in self._base_kernel.services.items():
            kernel.services[service_id] = service

        # Register plugins with current context
        # Mirrors: _kernel.Plugins.AddFromObject(new IntentPlugin(agentContext, ...))
        intent_plugin = IntentPlugin(context, kernel)
        kernel.add_plugin(intent_plugin, PluginNames.INTENT)

        m365_plugin = M365CopilotPlugin(context, kernel)
        kernel.add_plugin(m365_plugin, PluginNames.M365_COPILOT)

        openai_plugin = AzureOpenAIPlugin(context, kernel)
        kernel.add_plugin(openai_plugin, PluginNames.AZURE_OPENAI)

        synthesis_plugin = SynthesisPlugin(context, kernel)
        kernel.add_plugin(synthesis_plugin, PluginNames.SYNTHESIS)

        return kernel

    async def _analyze_intent(self, kernel: Kernel, query: str) -> List[Intent]:
        """
        Analyze user query to determine intents.

        Mirrors AnalyzeIntentAsync in OrchestratorAgent.cs:138-169.

        Args:
            kernel: Kernel with plugins registered.
            query: User query to analyze.

        Returns:
            List of detected intents with fallback to GeneralKnowledge.
        """
        try:
            # Invoke intent plugin (mirrors lines 140-146)
            result = await kernel.invoke(
                plugin_name=PluginNames.INTENT,
                function_name="AnalyzeIntent",
                query=query
            )

            json_response = str(result) if result else "[]"

            data = json.loads(json_response)
            intents = [Intent(**item) for item in data]

            return intents

        except Exception as e:
            self._logger.error(f"Error analyzing intent: {e}", exc_info=True)
            # Fallback to general knowledge on any error
            return [Intent(type=IntentType.GENERAL_KNOWLEDGE, query=query)]

    async def _execute_agents(
        self,
        kernel: Kernel,
        intents: List[Intent]
    ) -> List[AgentResponse]:
        """
        Execute agents based on detected intents.

        Mirrors ExecuteAgentsAsync in OrchestratorAgent.cs:172-193.

        Args:
            kernel: Kernel with plugins registered.
            intents: List of intents to process.

        Returns:
            List of agent responses.
        """
        if self._settings.orchestration.enable_parallel_execution:
            # Parallel execution with Task.WhenAll equivalent (lines 176-180)
            tasks = [
                self._execute_agent_for_intent(kernel, intent)
                for intent in intents
            ]
            responses = await asyncio.gather(*tasks, return_exceptions=True)

            # Convert exceptions to error responses
            result = []
            for i, response in enumerate(responses):
                if isinstance(response, Exception):
                    self._logger.error(
                        f"Error in parallel execution for intent {intents[i].type}: {response}"
                    )
                    result.append(AgentResponse(
                        agent=intents[i].type.value,
                        intent_type=intents[i].type,
                        content=f"Error: {str(response)}",
                        success=False
                    ))
                else:
                    result.append(response)

            return result
        else:
            # Sequential execution (lines 184-192)
            responses = []
            for intent in intents:
                response = await self._execute_agent_for_intent(kernel, intent)
                responses.append(response)
            return responses

    async def _execute_agent_for_intent(
        self,
        kernel: Kernel,
        intent: Intent
    ) -> AgentResponse:
        """
        Execute appropriate agent for a specific intent.

        Mirrors ExecuteAgentForIntentAsync in OrchestratorAgent.cs:195-228.

        Args:
            kernel: Kernel with plugins registered.
            intent: Intent to process.

        Returns:
            Agent response.
        """
        try:
            # Switch on intent type (mirrors lines 201-215)
            if intent.type == IntentType.M365_EMAIL:
                return await self._execute_m365_plugin(
                    kernel, "QueryEmails", intent.query, IntentType.M365_EMAIL
                )
            elif intent.type == IntentType.M365_CALENDAR:
                return await self._execute_m365_plugin(
                    kernel, "QueryCalendar", intent.query, IntentType.M365_CALENDAR
                )
            elif intent.type == IntentType.M365_FILES:
                return await self._execute_m365_plugin(
                    kernel, "QueryFiles", intent.query, IntentType.M365_FILES
                )
            elif intent.type == IntentType.M365_PEOPLE:
                return await self._execute_m365_plugin(
                    kernel, "QueryPeople", intent.query, IntentType.M365_PEOPLE
                )
            elif intent.type == IntentType.GENERAL_KNOWLEDGE:
                return await self._execute_general_knowledge(kernel, intent.query)
            else:
                return AgentResponse(
                    agent="unknown",
                    intent_type=intent.type,
                    content="I'm not sure how to handle that request.",
                    success=False
                )

        except Exception as e:
            # Mirrors lines 217-227
            self._logger.error(f"Error executing agent for intent {intent.type}: {e}")
            return AgentResponse(
                agent=intent.type.value,
                intent_type=intent.type,
                content=f"Error: {str(e)}",
                success=False,
                error=str(e)
            )

    async def _execute_m365_plugin(
        self,
        kernel: Kernel,
        function_name: str,
        query: str,
        intent_type: IntentType
    ) -> AgentResponse:
        """
        Execute M365 Copilot plugin function.

        Mirrors ExecuteM365PluginAsync in OrchestratorAgent.cs:230-265.

        Args:
            kernel: Kernel with plugins registered.
            function_name: Plugin function to invoke.
            query: Query to pass to function.
            intent_type: Type of intent being processed.

        Returns:
            Agent response.
        """
        result = await kernel.invoke(
            plugin_name=PluginNames.M365_COPILOT,
            function_name=function_name,
            query=query
        )

        content = str(result) if result else "No response received."

        return AgentResponse(
            agent="m365_copilot",
            intent_type=intent_type,
            content=content,
            success=True
        )

    async def _execute_general_knowledge(
        self,
        kernel: Kernel,
        query: str
    ) -> AgentResponse:
        """
        Execute Azure OpenAI plugin for general knowledge.

        Mirrors ExecuteGeneralKnowledgeAsync in OrchestratorAgent.cs:267-283.

        Args:
            kernel: Kernel with plugins registered.
            query: Query to answer.

        Returns:
            Agent response.
        """
        result = await kernel.invoke(
            plugin_name=PluginNames.AZURE_OPENAI,
            function_name="AnswerGeneralQuestion",
            question=query
        )

        content = str(result) if result else "I'm unable to answer that question."

        return AgentResponse(
            agent="azure_openai",
            intent_type=IntentType.GENERAL_KNOWLEDGE,
            content=content,
            success=True
        )

    async def _synthesize_response(
        self,
        kernel: Kernel,
        original_query: str,
        responses: List[AgentResponse]
    ) -> str:
        """
        Synthesize multiple agent responses into coherent answer.

        Mirrors SynthesizeResponseAsync in OrchestratorAgent.cs:285-306.

        Args:
            kernel: Kernel with plugins registered.
            original_query: Original user query.
            responses: List of agent responses to synthesize.

        Returns:
            Synthesized response.
        """
        # Format responses for synthesis
        synthesis_plugin = kernel.plugins[PluginNames.SYNTHESIS]
        formatted_responses = SynthesisPlugin.format_responses_for_synthesis(responses)

        # Invoke synthesis
        result = await kernel.invoke(
            plugin_name=PluginNames.SYNTHESIS,
            function_name="Synthesize",
            original_query=original_query,
            responses=formatted_responses
        )

        return str(result) if result else "I couldn't generate a response."


# Factory function to create and configure the orchestrator
def create_orchestrator_agent(settings: AppSettings, connection_manager: MsalConnectionManager) -> AgentApplication[TurnState]:
    """
    Create and configure the orchestrator agent application.

    This is the main entry point that should be used in main.py.

    Args:
        settings: Application configuration.
        connection_manager: MsalConnectionManager instance.

    Returns:
        Configured AgentApplication instance ready for use with CloudAdapter.
    """
    orchestrator = OrchestratorAgentApp(settings=settings, connection_manager=connection_manager)
    return orchestrator.agent_app
